### ISSUE TITLE
[firebase_auth] on web authStateChanges sink null value then correct authState on page reload

### ISSUE THREAD

**Issue by D-Jeremie** [firebase/flutterfire#3356]
**Describe the bug**
On web, reloading a page with a previously signed in user makes ```authStateChanges()``` sink two states one after the other.
The first one is ```null``` and the other one is the correct one, with the real authState of the previously signed in user.
In android ```authStateChanges()```also sink two states but both are the real authState of the user.

This make weird behavior if UI is linked with authStateChanges values.

**To Reproduce**
Create à project with Firebase Auth enabled
Then run 
```
import 'dart:async';

import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp();

  runApp(MaterialApp(
    title: 'Example App',
    theme: ThemeData.dark(),
    home: AuthExample(),
  ));
}

/// Provides a UI to authenticate
class AuthExample extends StatefulWidget {
  @override
  _AuthExampleState createState() => _AuthExampleState();
}

class _AuthExampleState extends State<AuthExample> {
  String _log = "";

  ///TODO enter valid email and password
  String email = "userEmailToLogIn@mail.com";
  String password = "userPassword";

  @override
  void initState() {
    if (kIsWeb) FirebaseAuth.instance.setPersistence(Persistence.LOCAL);
    int i = 0;

    ///Listen on authStateChanges for log
    FirebaseAuth.instance.authStateChanges().listen((event) {
      final DateTime time = DateTime.now();
      print("timestamp : $time");
      print("Event n°$i : $event");
      _log = "timestamp: $time\nEvent n°$i : ${event?.uid}\n******\n$_log";
      if (mounted) {
        setState(() {});
      }
      i++;
    });

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Firebase Example App"),
      ),
      body: Column(
        children: [
          Align(
            alignment: Alignment.centerLeft,
            child: Padding(
              padding: const EdgeInsets.all(8.0),
              child: Text(
                "Log :",
                style: Theme.of(context).textTheme.headline6,
              ),
            ),
          ),
          SizedBox(
              height: 500, child: SingleChildScrollView(child: Text(_log))),
          StreamBuilder<User>(
            stream: FirebaseAuth.instance.authStateChanges(),
            builder: (context, snapshot) {
              if (snapshot.connectionState == ConnectionState.waiting) {
                return const Center(
                  child: CircularProgressIndicator(),
                );
              } else {
                if (snapshot.hasError) {
                  return Center(
                    child: Text(snapshot.error.toString()),
                  );
                } else {
                  return Container(
                    ///Add color to better see the weird UI reload that it cause
                    color: (snapshot.hasData && snapshot.data.uid != null)
                        ? Colors.transparent
                        : Colors.white,
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: <Widget>[
                        if (snapshot.hasData && snapshot.data.uid != null)
                          Container(
                            padding: const EdgeInsets.all(16),
                            alignment: Alignment.center,
                            child: FlatButton.icon(
                              icon: const Icon(Icons.exit_to_app),
                              color: Colors.red,
                              label: const Text('Sign Out'),
                              onPressed: () {
                                print("Signing out");
                                FirebaseAuth.instance.signOut();
                              },
                            ),
                          )
                        else
                          Container(
                            padding: const EdgeInsets.all(16),
                            alignment: Alignment.center,
                            child: FlatButton.icon(
                              icon: const Icon(Icons.verified_user),
                              color: Colors.orange,
                              label: const Text('Sign In'),
                              onPressed: () {
                                print("Signing in");
                                FirebaseAuth.instance
                                    .signInWithEmailAndPassword(
                                        email: email, password: password)
                                    .then((value) => print(value),
                                        onError: (error) {
                                  Scaffold.of(context).showSnackBar(SnackBar(
                                      content:
                                          Text("Error : ${error.toString()}")));
                                  print("Error : $error");
                                });
                              },
                            ),
                          ),
                      ],
                    ),
                  );
                }
              }
            },
          ),
        ],
      ),
    );
  }
}

```
**Expected behavior**
Same as android, sink the right authState at first time.
Better if it's think only one authState instead of two.

**Additional context**
Using ```firebase_auth: ^0.18.0+1```
Result of a page reload on a previously logged in user : 
Event 0 : null
Event 1 : the correct user id loaded
![image](https://user-images.githubusercontent.com/33086704/91286031-38a60c80-e78e-11ea-8437-3429cd58a07a.png)

**Flutter doctor**
```
[✓] Flutter (Channel beta, 1.20.2, on Mac OS X 10.15.6 19G2021, locale fr-FR)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
[✓] Chrome - develop for the web
[✓] Android Studio (version 3.6)
[✓] VS Code (version 1.48.1)
[✓] Connected device (3 available)

• No issues found!
```

**Comment by edufolly** [comment]
I have the same problem.

```
[✓] Flutter (Channel beta, 1.21.0-9.1.pre, on Mac OS X 10.15.4 19E287, locale pt-BR)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.4)
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.0)
[✓] IntelliJ IDEA Community Edition (version 2020.2)
[✓] VS Code (version 1.48.2)
[✓] Connected device (2 available)

• No issues found!
```

I've been creating a timeout on authStateChanges stream and catch the error to reflect the user is not logged in.

```
FirebaseAuth.instance.authStateChanges().timeout(Duration(seconds: 2))
```

In this case, I ignored the null values and user snapshot.hasError to detect no user logged.

**Comment by lucashilles** [comment]
The [documentation](https://firebase.google.com/docs/auth/web/manage-users#get_the_currently_signed-in_user) of firebase web sdk contains something about this.

_Note: currentUser might also be null because the auth object has not finished initializing. If you use an observer to keep track of the user's sign-in status, you don't need to handle this case._

**Comment by Ehesp** [comment]
Yeah this is something we need to address once we focus on web, it's currently on handled on native only. Basically what @lucashilles is correct, what we need to be doing is setting up a initialization flow when `initializeApp` is called (which is what happens on native) which allows us to fetch any constant/pre-defined values before you start using the app. In this case it'd be something like:

Unfortunately this isn't so simple due to a number of issues (has to be done pre-registration, handling errors (e.g. trying to access firebase if it's not been included in the `index.html` file etc)).

**Comment by Ehesp** [comment]
Not tested, but a stinky hack could be doing this:

```
// Import/add the firebase-dart lib
import 'package:firebase/firebase.dart' as firebase;


Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp();

  // Force auth initialization
  await firebase
    .auth(firebase.app(app.name))
    .onAuthStateChanged
    .first;

  runApp(MaterialApp(
    title: 'Example App',
    theme: ThemeData.dark(),
    home: AuthExample(),
  ));
}
```

That might work for now.

**Comment by buivanthu** [comment]
I have same problem :(

**Comment by sjones512** [comment]
This issue was bugging me more than it should have.  I'm not sure if it is how I am consuming the authStateChanges, but @Ehesp's fix didn't seem to work for me.  What I ended up doing is trying to "warm up" the authStateChanges from `index.html`

```
firebase.initializeApp(firebaseConfig);
const unsub = firebase.auth().onAuthStateChanged(() => unsub());
```

**Comment by edmbn** [comment]
I was able to solve it by using @sjones512 proposed solution. This is still an issue to be looked at.

**Comment by krisztianodor** [comment]
Yes, all these streams are running two times on ios and android too at startup if the user is previously signed in:

- authStateChanges()
- userChanges()
- idTokenChanges()

Each runs twice, both times with the same correct value - at least on ios and android.

_I know that userChanges() stream is a superset of both [authStateChanges] and [idTokenChanges], i just mention that each runs twice._

For me this is not a big problem right now, I just have an unnecessary bloc state update. But with a minor workaround i can solve this 'problem', by simply checking if the value is the same. Hopefully this will be fixed soon.

**Comment by mtkgeek** [comment]
This can be fixed by using dart async's `distinct` method.
For cases where both values returned are the same i.e neither is null
`await firebase
    .auth(firebase.app(app.name))
    .authStateChanges()
          .distinct((first, second) => first.uid == second.uid);
`

And if the first value returned is null
`await firebase
    .auth(firebase.app(app.name))
    .authStateChanges()
          .distinct((first, _) => first.uid == null);
`

**Comment by reovo** [comment]
I'm facing similar issue on iOS, then trying to `signOut()`
`authStateChanges()` is still returning not null value. 

I noticed such behaviour only for some accounts. Cannot reproduce that on simulator.  
The issue only on real device and only when i install via TestFlight.

**Comment by devj3ns** [comment]
I have the same problem (using Flutter to export for Web).

These are my imports in the index.html file:
```
<script src="https://www.gstatic.com/firebasejs/8.1.1/firebase-app.js"></script>
<script src="https://www.gstatic.com/firebasejs/8.1.1/firebase-auth.js"></script>
```

When I change them to version 7.5.0 it works fine (all newer versions have the issue):
```
<script src="https://www.gstatic.com/firebasejs/7.5.0/firebase-app.js"></script>
<script src="https://www.gstatic.com/firebasejs/7.5.0/firebase-auth.js"></script>
```

**Comment by zibizz1** [comment]
Confirmed.
On js SDK 8.0.1 when refreshing the page user is not auto-logged in.
On 7.5.0 everything is OK

**Comment by Sun3** [comment]
Is there an update, I am also having this issue.

Thanks.

**Comment by azorychta** [comment]
> Not tested, but a workaround could be doing this:
> 
> ```dart
> // Import/add the firebase-dart lib
> import 'package:firebase/firebase.dart' as firebase;
> 
> 
> Future<void> main() async {
>   WidgetsFlutterBinding.ensureInitialized();
>   await Firebase.initializeApp();
> 
>   // Force auth initialization
>   await firebase
>     .auth(firebase.app(app.name))
>     .onAuthStateChanged
>     .first;
> 
>   runApp(MaterialApp(
>     title: 'Example App',
>     theme: ThemeData.dark(),
>     home: AuthExample(),
>   ));
> }
> ```
> 
> That might work for now.

I tried everything I could for 5 hours and this is the only solution that works. Except, I couldn't figure out how to include an app name, so I got rid of it and it still works:

 // Force auth initialization
   await firebase
     .auth(firebase.app())
     .onAuthStateChanged
     .first;

Happy coding!!! And Flutter, please fix this.

**Comment by azorychta** [comment]
I’ve updated to that version and beyond (0.18.4+1) and it still doesn’t work to fix the issue. :(

What is the code from firebase-dart doing on web that isn’t done by FirebaseAuth.instance.authStateChanges().first ?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by zibizz1** [comment]
I listen also for userChanges, and that solved the issue. Propably there was change in authStateChanges.
`_firebaseAuth.authStateChanges().listen((event) {
      if (event != null) {
        _loadUser(event);
      }
    });
    _firebaseAuth.userChanges().listen((event) {
      if (event != null) {
        _loadUser(event);
      }
    });`
    
    and last compatible version for flutter web is
`<script src="https://www.gstatic.com/firebasejs/7.20.0/firebase-app.js"></script>
    <script src="https://www.gstatic.com/firebasejs/7.20.0/firebase-analytics.js"></script>`

**Comment by azorychta** [comment]
A thousand thank yous, zibizz1!
I was using a StreamBuilder for authentication and for me I used your help to just change authStateChanges() to userChanges() as the stream to listen to. It worked!

Also a note for those sticking to 7.2.0 to get this to work on web: you need to make sure Firestore is at 7.21.0 to keep the filtered queries that use !=

________________________________

**Comment by Sun3** [comment]
@azorychta If I understand correctly, you are no longer using StreamBuilder to watch for Authentication changes and are using the listen events instead?

**Comment by azorychta** [comment]
In my StreamBuilder I simply changed “stream: authService.authStateChanges()” to “stream: authService.userChanges()” – hope that helps.

**Comment by deepakrana0011** [comment]
> This issue was bugging me more than it should have. I'm not sure if it is how I am consuming the authStateChanges, but @Ehesp's fix didn't seem to work for me. What I ended up doing is trying to "warm up" the authStateChanges from `index.html`
> 
> ```
> firebase.initializeApp(firebaseConfig);
> const unsub = firebase.auth().onAuthStateChanged(() => unsub());
> ```

@sjones512  its working fine on locally,  but when i am deploying it ,its not working. Have any idea?

**Comment by cedvdb** [comment]
I'm skipping the first value, 

that doesn't seem really future proof though. Not tested on mobile.

```
    return _authFire
        .authStateChanges()
        .asBroadcastStream()
        // currently firebase auth fires twice,
        // when the user is initially authenticated the first value
        // is null, ence why we skip it
        .skip(1)
        .map((user) => AuthUser(id: user?.uid));
```

**Comment by ln-12** [comment]
I got the following solution working on web as `StreamBuilder` and `FutureBuilder` didn't update the state correctly:
```
   int state = -1;

  @override
  void initState() {
    super.initState();

    FirebaseAuth.instance.authStateChanges().listen((User? user) {
      if (user == null) {
        setState(() {
          state = 0;
        });
      } else {
        setState(() {
          state = 1;
        });
      }
    });
  }

  Widget getBody() {
    if (state == 0) {
      return LoginForm();
    } else if (state == 1) {
      return HomeScreen();
    }

    return CircularProgressIndicator();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body:
          Center(child: Padding(padding: EdgeInsets.all(8), child: getBody())),
    );
  }
}
```

Maybe it helps someone :)

**Comment by NamanShergill** [comment]
This is a major bug that messes up an app's startup logic, how come it still has not been worked on for an year?

**Comment by omatt** [comment]
> @sjones512 its working fine on locally, but when I am deploying it ,its not working. Have any idea?

The issue occurs on page refresh. In my experience, running the app on web using `flutter run` to override the currently running web app retains the auth state. But once the page is reloaded/refreshed, the auth state is lost.

**Comment by LucaDillenburg** [comment]
None of the fixes above helped. As I understand it, every time you reload the page Flutter Web looses the auth state. Is this still the case? How is this major issue not addressed even after more than one year?

**Comment by cyberpwnn** [comment]
Why is this still broken? What are all the package updates even changing? Is this a joke?

**Comment by aditya-nahak** [comment]
> > This issue was bugging me more than it should have. I'm not sure if it is how I am consuming the authStateChanges, but @Ehesp's fix didn't seem to work for me. What I ended up doing is trying to "warm up" the authStateChanges from `index.html`
> > ```
> > firebase.initializeApp(firebaseConfig);
> > const unsub = firebase.auth().onAuthStateChanged(() => unsub());
> > ```
> 
> @sjones512 its working fine on locally, but when i am deploying it ,its not working. Have any idea?

Did you find any solution for this ?

**Comment by tijanirf** [comment]
The `Persistence` is somehow broken on the latest version, I think the fix is on the way: https://github.com/firebase/flutterfire/pull/9138

**Comment by sunnyp18** [comment]
I've been scratching my eyes out trying to find a solution to this! Any workarounds? 

Will sharedpreferences solve this?

**Comment by RajendraAVerma** [comment]
This issue is fixed Now.  #9089 
This fix has just been released in the latest version 3.6.0. Thanks.

**Comment by russellwheatley** [comment]
Hey folks, I've tested this on the web platform, and it no longer passes a `null` value on page refresh or hot restart if the user has already signed in for `authStateChanges()` 🙏
